### ISSUE TITLE
Section substitution does not work together with factors that are not an environment

### ISSUE THREAD

**Issue by pytoxbot** [tox-dev/tox#312]
- Bitbucket: https://bitbucket.org/hpk42/tox/issue/312
- Originally reported by: @JulianB
- Originally created at: 2016-02-10T01:19:59.384

Reproducing tox.ini:

```
#!ini

[tox]
envlist = pypy, coverage

[testenv]
deps =
    pypy,coverage: twisted

[testenv:coverage]
deps =
    {[testenv]deps}
    coverage
```

which attempts to tell pip to install a req called `pypy,coverage: twisted` when the coverage test environment is run with `tox -e coverage`.

(The `pypy` environment works fine).

**Comment by obestwalter** [comment]
This looks like an ordering problem: the factors need to be resolved before the section substitution. I would like to say this might be easy to fix, but the code handling the configuration has been growing for a while without any refactorings and is sufficiently convoluted. Definitely fixable though.

**Comment by obestwalter** [comment]
see #509 for more interesting details :)

**Comment by jayvdb** [comment]
@pytoxbot, could you re-test.

I believe this was fixed in 3.5.3, very likely in https://github.com/tox-dev/tox/pull/1058
@gaborbernat
